### PR TITLE
update optimistic-transaction.md

### LINKED ISSUE #114: Write down the first version of Optimistic Transaction
https://github.com/pingcap/tidb-dev-guide/blob/master/src/understand-tidb/optimistic-transaction.md
cc @lysu @tisonkun

### PR DESCRIPTION
**Pull request by TonsnakeLin**
### What issue does this PR solve?

- close #114 

### What is changed:
Added dev guide article for optimistic transaction

### CODE REVIEW

**Comment by you06** [comment]
There are still some format issues:
- The bold marks are not work in code blocks, which should be removed.
- The main body's function names are in quotation marks, it would be better to put them in backquotes, so they look like `func fn()`.

For the contents of this patch, there are many details of how TiDB implement optimistic transaction like batch processing which looks good, however, some details are missing like error handling(retry or resolve locks).

I think we can accept this document after the format issues are fixed. If someone thinks there are too many code references or any details that are worth mentioned, a further patch will improve this document.

**Comment by TonsnakeLin** [comment]
> There are still some format issues:
> 
> * The bold marks are not work in code blocks, which should be removed.
> * The main body's function names are in quotation marks, it would be better to put them in backquotes, so they look like `func fn()`.
> 
> For the contents of this patch, there are many details of how TiDB implement optimistic transaction like batch processing which looks good, however, some details are missing like error handling(retry or resolve locks).
> 
> I think we can accept this document after the format issues are fixed. If someone thinks there are too many code references or any details that are worth mentioned, a further patch will improve this document.

I have fixed the format issues, thanks. Why so mush code in this paper? I think the new developers want to position the important functions quickly and read the code in detail for a transaction process flow.

**Comment by TonsnakeLin** [comment]
@you06 @cfzjywxk  please review or approve, thanks

File: src/understand-tidb/optimistic-transaction.md [hunk 1]
```diff
@@ -1 +1,490 @@
 # Optimistic Transaction
+
+Under the optimistic transaction model, modification conflicts are regarded as part of the transaction commit. TiDB cluster uses the pessimistic transaction model by default before version 3.0.8, uses pessimistic transaction model after version 3.0.8. System variable tidb_txn_mode controls TiDB uses optimistic transaction mode or pessimistic transaction mode.
```
Comment (you06) [inline src/understand-tidb/optimistic-transaction.md:3]:
```suggestion
Under the optimistic transaction model, modification conflicts are regarded as part of the transaction commit. TiDB cluster uses the optimistic transaction model by default before version 3.0.8, uses pessimistic transaction model after version 3.0.8. System variable tidb_txn_mode controls TiDB uses optimistic transaction mode or pessimistic transaction mode.
```

File: src/understand-tidb/optimistic-transaction.md [hunk 2]
```diff
@@ -1 +1,490 @@
...
+​            (e *UpdateExec) updateRows
+​                 Next
+​                   (e *PointGetExecutor) Next
+```
+
+
+
+### (e *UpdateExec) updateRows
+
+The function `(e *UpdateExec) updateRows` does the main work for update statement. he important comment and simplified code is as followers. The completed code is [here](https://github.com/pingcap/tidb/blob/cd8fb24c5f7ebd9d479ed228bb41848bd5e97445/executor/update.go#L229) .
```
Comment (you06) [inline src/understand-tidb/optimistic-transaction.md:108]:
```suggestion
The function `(e *UpdateExec) updateRows` does the main work for update statement. The important comment and simplified code are as followers. The completed code is [here](https://github.com/pingcap/tidb/blob/cd8fb24c5f7ebd9d479ed228bb41848bd5e97445/executor/update.go#L229) .
```

File: src/understand-tidb/optimistic-transaction.md [hunk 3]
```diff
@@ -1 +1,490 @@
...
+​        return c.commitTxn(ctx, commitDetail)
+​    } 
+}
+```
+
+
+
+### prewrite
+
+The entry function to prewrite a transaction is `(c *twoPhaseCommitter) prewriteMutations` which calls the function `(batchExe *batchExecutor) process` to do it. the function `(batchExe *batchExecutor) process` calls `(batchExe *batchExecutor) startWorker` to prewrite evenry batch parallelly. The function  `(batchExe *batchExecutor) startWorker` calls `(action actionPrewrite) handleSingleBatch` to prewrite a single batch.
```
Comment (you06) [inline src/understand-tidb/optimistic-transaction.md:200]:
```suggestion
The entry function to prewrite a transaction is `(c *twoPhaseCommitter) prewriteMutations` which calls the function `(batchExe *batchExecutor) process` to do it. The function `(batchExe *batchExecutor) process` calls `(batchExe *batchExecutor) startWorker` to prewrite evenry batch parallelly. The function  `(batchExe *batchExecutor) startWorker` calls `(action actionPrewrite) handleSingleBatch` to prewrite a single batch.
```

File: src/understand-tidb/optimistic-transaction.md [hunk 4]
```diff
@@ -1 +1,490 @@
...
+​              return errors.Trace(err) // backoff exceeded maxtime, returns error
+​           }
+​       }
+​    }// for loop
+}
+```
+
+### commit
+
+The entry function to commit a transaction is `(c *twoPhaseCommitter) commitMutations` which calls the function `(c *twoPhaseCommitter) doActionOnGroupMutations` to do it. the function `(batchExe *batchExecutor) process` calls `(batchExe *batchExecutor) startWorker` to prewrite evenry batch parallelly. The function  `(batchExe *batchExecutor) startWorker` calls `(actionCommit) handleSingleBatch` to commit a single batch.
```
Comment (you06) [inline src/understand-tidb/optimistic-transaction.md:345]:
```suggestion
The entry function of commiting a transaction is `(c *twoPhaseCommitter) commitMutations` which calls the function `(c *twoPhaseCommitter) doActionOnGroupMutations` to do it. The batch of primary key will be committed first, then the function `(batchExe *batchExecutor) process` calls `(batchExe *batchExecutor) startWorker` to commit the rest batches parallelly and asynchronously. The function  `(batchExe *batchExecutor) startWorker` calls `(actionCommit) handleSingleBatch` to commit a single batch.
```

File: src/understand-tidb/optimistic-transaction.md [hunk 5]
```diff
@@ -1 +1,490 @@
...
+
+The important comment and simplified code of function `(batchExe *batchExecutor) process`  are as mentioned above in prewrite part . The completed code is [here](https://github.com/tikv/client-go/blob/843a5378aa9101c0e2aba61e0c2c11b3f122f08f/txnkv/transaction/2pc.go) .
+
+
+
+#### (actionCommit) handleSingleBatch
+
+The function `(batchExe *batchExecutor) process` calls the function `(actionCommit) handleSingleBatch` to send commit request to all tikv nodes.
+
+The important comment and simplified code are as followers. The completed code is [here](https://github.com/tikv/client-go/blob/843a5378aa9101c0e2aba61e0c2c11b3f122f08f/txnkv/transaction/2pc.go) .
```
Comment (you06) [inline src/understand-tidb/optimistic-transaction.md:403]:
```suggestion
The important comment and simplified code are as followers. The completed code is [here](https://github.com/tikv/client-go/blob/843a5378aa9101c0e2aba61e0c2c11b3f122f08f/txnkv/transaction/commit.go#L67) .
```

File: src/understand-tidb/optimistic-transaction.md [hunk 6]
```diff
@@ -1 +1,490 @@
...
+​    }// for loop
+}
+```
+
+
+
+
+
+
+
```
Comment (you06) [inline src/understand-tidb/optimistic-transaction.md:490]: One empty line is enough, may remove the following lines.

**Comment by you06** [comment]
rest LGTM

File: src/understand-tidb/optimistic-transaction.md [hunk 7]
```diff
@@ -1 +1,490 @@
...
```
Comment (zhangyangyu) [inline src/understand-tidb/optimistic-transaction.md:490]:
```suggestion
```

**Pull request merged**